navigation/navigation-compose-material/src/main/java/androidx/navigation/compose/material/BottomSheetNavigator.kt

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright 2021 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package androidx.navigation.compose.material
+
+import androidx.compose.foundation.layout.ColumnScope
+import androidx.compose.material.ExperimentalMaterialApi
+import androidx.compose.material.ModalBottomSheetState
+import androidx.compose.material.ModalBottomSheetValue
+import androidx.compose.material.rememberModalBottomSheetState
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.collectAsState
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.saveable.rememberSaveableStateHolder
+import androidx.compose.runtime.setValue
+import androidx.lifecycle.Lifecycle
+import androidx.navigation.FloatingWindow
+import androidx.navigation.NavBackStackEntry
+import androidx.navigation.NavController
+import androidx.navigation.NavDestination
+import androidx.navigation.Navigator
+import androidx.navigation.NavigatorState
+import androidx.navigation.compose.material.BottomSheetNavigator.Destination
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+
+/**
+ * Create and remember a [BottomSheetNavigator]
+ *
+ * @param sheetState The [ModalBottomSheetState] that the [BottomSheetNavigator] will use to
+ * drive the sheet state
+ * @param navController The [NavController] used to the pop the back stack when the sheet is
+ * dismissed. Please note that this will be redundant in the next release.
+ */
+@OptIn(ExperimentalMaterialApi::class)
+@Composable
+public fun rememberBottomSheetNavigator(
+    sheetState: ModalBottomSheetState =
+        rememberModalBottomSheetState(ModalBottomSheetValue.Hidden),
```

Comment:
These lines brought up an interesting question I had:
- We need to allow hoisting this state to allow for coordinating other things with the bottom sheet, is that correct? Or is there some other reason?
- What happens when someone calls `hide()` or `show()` on this state externally from the `BottomSheetNavigator`? What does that do to the `BottomSheetNavigator`?
- What happens after a configuration change? With an initial state of always `ModalBottomSheetValue.Hidden`, does that mean that when the `NavController` restores its state that the bottom sheet animates in again? I guess we don't know the 'intial state' until the `BottomSheetNavigator` has its `onAttach()` called and we can check the `NavigatorState.backStack` to see if it empty or not? I don't see any 'snap to state' kind of API?

Reply:
Great question!

1. Yes! There might be cases where you'd want to hide or show the sheet (though I imagine most of these use cases would be covered by simply navigating) manually. Theoretically, we do not need to offer this control (it would save us some headache when thinking about the initial bottom sheet state; what if somebody set it to `Expanded` here?) though. I'd be open to removing this param! It would help prevent misuse and the `sheetState` can be accessed through the navigator in any case. I'll play around with that a little bit :)
2. The only way we can react to `show`/`hide` calls is by looking at the `sheetVisiblity` - we currently have a `snapshotFlow` observing the visibility and calling a callback (that will then pop the back stack) when the sheet was visible and changed its state to hidden. So that `LaunchedEffect`/`snapshotFlow` covers both manual and programmatic dismiss (we have tests for that in `SheetContentHostTest` too!).  `show()`ing the sheet is a good question... I thought about it but didn't find any good solution. Calling `show()` is a little ambiguous: 
    1. It implies the sheet is currently hidden and the current nav destination is not a `BottomSheetNavigator.Destination`
    2. When you call `show()`, what is the expected outcome? What should we show?
Do you think we should just ignore `show` calls? Or any other ideas?
3. `rememberModalBottomSheetState` uses `rememberSaveable` under the hood so we will stay in the same state across config changes. But we might not actually want that as it means that the sheet and navigator might be out of sync. I think it'd be better to react to that after the config change though, when we actually know if we want to show the sheet or not.
> snap to state API
In the navigator?

Reply:
(also, if you're asking these questions, it makes sense to document that - I'll add explanations around this to the docs later on!)

Reply:
Sooo
1. I just went back and forth on this - IMO we should keep exposing this because folks might want to observe the state to run things etc. Maybe a point can be made that this should be done using effects so I'm not entirely against removing it and adding it back if requested.
2. I'm still not sure what we should do when somebody calls `show` - we could just ignore it. Feels like a sensible thing to do to me. Wouldn't be an issue if we weren't hoisting the state.
3. Is working now 🎉 Will add a test